smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/impl/ReactiveKafkaProducer.java

```diff
@@ -147,18 +159,32 @@ public Uni<Void> runOnSendingThread(java.util.function.Consumer<Producer<K, V>>
     @CheckReturnValue
     public Uni<RecordMetadata> send(ProducerRecord<K, V> record) {
         return withProducerOnSendingThread()
-                .chain(c -> Uni.createFrom().emitter(em -> c.send(record, (metadata, exception) -> {
-                    if (exception != null) {
-                        if (record.topic() != null) {
-                            log.unableToWrite(this.channel, record.topic(), exception);
-                        } else {
-                            log.unableToWrite(this.channel, exception);
-                        }
-                        em.fail(exception);
+                .chain(c -> {
+                    final ProducerRecord<K, V> intercepted;
+                    if (interceptor != null) {
+                        intercepted = interceptor.onSend(record);
                     } else {
-                        em.complete(metadata);
+                        intercepted = record;
                     }
-                })));
+                    return Uni.createFrom().emitter(em -> c.send(intercepted, (metadata, exception) -> {
+                        if (exception != null) {
+                            if (interceptor != null) {
+                                interceptor.onAcknowledgement(getRecordMetadataForFailure(intercepted), exception);
```

Comment:
Wondering if we need a onNegativeAcknowledgement method instead (to be clear)

Reply:
The idea is to be able to use the standard ProducerInterceptor interface from the Kafka package.